test/unit/Monitoring/OperationTimerTest.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * @author    Craig Gosman <craig@ingenerator.com>
+ * @licence   proprietary
+ */
+
+
+namespace test\unit\Ingenerator\PHPUtils\unit\Monitoring;
+
+
+use DateTimeImmutable;
+use Ingenerator\PHPUtils\DateTime\Clock\RealtimeClock;
+use Ingenerator\PHPUtils\DateTime\Clock\StoppedMockClock;
+use Ingenerator\PHPUtils\Monitoring\ArrayMetricsAgent;
+use Ingenerator\PHPUtils\Monitoring\MetricId;
+use Ingenerator\PHPUtils\Monitoring\OperationTimer;
+use PHPUnit\Framework\TestCase;
+
+class OperationTimerTest extends TestCase
+{
+    protected ArrayMetricsAgent $metrics;
+    protected ?RealtimeClock $real_time_clock = NULL;
+
+    public function test_it_is_initialisable()
+    {
+        $this->assertInstanceOf(OperationTimer::class, $this->newSubject());
+    }
+
+    /**
+     * @testWith [false]
+     *           [true]
+     *           ["I'm a string"]
+     *
+     * @param mixed $callback
+     */
+    public function test_timeOperation_calls_callback($callback)
+    {
+        $subject = $this->newSubject();
+        $this->assertSame($callback, $subject->timeOperation(function () use ($callback) { return $callback; }));
+    }
+
+    public function test_timeOperation_creates_metric()
+    {
+        $subject = $this->newSubject();
+        $metric  = $subject->timeOperation(
+            function (MetricId $metric) {
+                $this->assertMetricMatches(['name' => 'queries', 'source' => 'mysql'], $metric);
+                $metric->setName('queries.failed');
+                $metric->setSource('mysql.slave');
+
+                return $metric;
+            },
+            'queries',
+            'mysql'
+        );
+        $this->assertMetricMatches(['name' => 'queries.failed', 'source' => 'mysql.slave'], $metric);
+    }
+
+    public function test_timeOperation_times_operation()
+    {
+        $this->real_time_clock = StoppedMockClock::at('10:00:00');
+        $subject               = $this->newSubject();
+        $clock                 = $this->real_time_clock;
+        $subject->timeOperation(function () use ($clock) { $clock->tickMicroseconds(250); }, 'test');
+
+        $metric = $this->metrics->getTimers()[0];
+        $start  = $metric['start'];
+        /** @var $start DateTimeImmutable */
+        $end = $metric['end'];
+        /** @var $end DateTimeImmutable */
+        $this->assertEquals('10:00:00.000000', $start->format('H:i:s.u'));
+        $this->assertEquals('10:00:00.000250', $end->format('H:i:s.u'));
+    }
+
```

Comment:
Also need a test that the metric is captured on exception / throwable, maybe:
```
function test_times_operation_on_child_exception() {
  // setup as above
  $e = new \InvalidArgumentException('testing');
  try {
    $subject->timeOperation(function () use ($clock, $e) { 
      $clock->tickMicroseconds(500);
      throw $e;
    }
   )
  } catch (InvalidArgumentException $got_e) {
    $this->assertSame($e, $got_e, 'sanity check in case of something else');
  }
  // assert the captured metric timing
}
```

Could extract a method for actually asserting the start / end times.